node/src/utils/pidfile.rs

```diff
@@ -0,0 +1,255 @@
+//! Pidfile utility type.
+//!
+//! Pidfiles are used to gate access to a resource, as well as detect unclean shutdowns.
+
+use std::{
+    fs::{self, File},
+    io::{self, Read, Seek, SeekFrom, Write},
+    num::ParseIntError,
+    path::{Path, PathBuf},
+    process,
+};
+
+use fs2::FileExt;
+use thiserror::Error;
+use tracing::warn;
+
+/// A PID file.
+///
+/// Records the current process' PID, removes it on exit. Can be used to determine whether or not
+/// an application was shut down cleanly.
+///
+/// The pidfile is held open with an exclusive but advisory lock.
+#[derive(Debug)]
+pub struct Pidfile {
+    /// The pidfile.
+    ///
+    /// The file will be locked for the lifetime of `Pidfile`.
+    pidfile: File,
+    /// The pidfile location.
+    path: PathBuf,
+    /// Previous pidfile contents.
+    previous: Option<u32>,
+}
+
+/// An error acquiring a pidfile.
+#[derive(Debug, Error)]
+pub enum PidfileError {
+    /// The pidfile could not be opened at all.
+    #[error("could not pidfile: {0}")]
```

Comment:
```suggestion
    #[error("could not open pid file: {0}")]
```

I also think all instances of "pidfile" should be changed to "pid file"